Added maps with walking directions overlay

notthatmj opened the pull request:
While adding the maps, I also renamed `BusinessDataController` to just `Model`, and moved the setup for the `Model` instance into the AppDelegate class.

notthatmj committed:
Change navigation bar text

notthatmj committed:
Add README. Fix test.

notthatmj committed:
Remove unreliable UI test

notthatmj committed:
Remove unnecessary scheme and edit README

notthatmj committed:
Edit README

notthatmj committed:
Fix bug: activity indicator not showing on startup

notthatmj committed:
Merge branch 'hotfix/v0.1.1' into development

notthatmj committed:
Add MapViewController

notthatmj committed:
Add property

notthatmj committed:
Zoom in on an arbitrary location

notthatmj committed:
Zoom to business location

notthatmj committed:
Add back button to map scene

notthatmj committed:
Add pin for business

notthatmj committed:
Show business name and user location on map

notthatmj committed:
Pan to user location instead of business location

notthatmj committed:
Set zoom radius based on distance

notthatmj committed:
Rename property

notthatmj committed:
Display walking directions

notthatmj committed:
Merge branch 'feature-maps' into development

jacobvanorder commented on Business Finder/Model.h:
```diff
@@ -0,0 +1,28 @@
+
+#import <Foundation/Foundation.h>
+#import "LocationGateway.h"
+#import "FourSquareGateway.h"
```
You shouldn't be importing `.h` files into another `.h` file unless you declare a protocol on that `.h` file that you must use. Instead, `@class FourSquareGateway;`

notthatmj replied:
I need to import ""LocationGateway.h" and "FourSquareGateway.h" because `Model` implements the protocols `LocationGatewayDelegate` and `FourSquareGatewayDelegate`.

jacobvanorder commented on Business Finder/AppDelegate.h:
```diff
@@ -8,10 +8,11 @@
 
 @property (strong, nonatomic) UIWindow *window;
-
+@property (strong, nonatomic) Model *model;
```
This is a model object that holds state in the AppDelegate, correct? If so, that is a little frown upon within puritanical types. Ways around it? Save that state to disk and retrieve independently throughout the app or pass the state around.

notthatmj replied:
I removed the `model` property from the `AppDelegate`, and refactored things so we just pass around the relevant data between view controllers.

jacobvanorder commented on Business Finder/BusinessCell.h:
```diff
@@ -0,0 +1,14 @@
+#import "Business.h"
+
+@interface BusinessCell : UITableViewCell
+@property (strong, nonatomic) Business *business;
```
Having a model be a property of a view is not MVC.

notthatmj replied:
Good point. Now the cell just stores the index, and `NearbyBusinessesTableViewController-prepareForSegue:` retrieves the appropriate `Business` instance and uses it to configure the destination `MapViewController`

jacobvanorder commented on Business Finder/NearbyBusinessesTableViewController.m:
```diff
@@ -23,7 +25,30 @@ - (void)viewDidLoad {
+-(void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
+    MapViewController *mapViewController = segue.destinationViewController;
+    BusinessCell *cell = sender;
+    mapViewController.business = cell.business;
```
Here, you should have an array of all of the businesses and then get to it using the cell's index path in your tableview.

notthatmj replied:
I've modified the code to basically do this, but instead of directly accessing an array, I delegate retrieval through `NearbyBusinessesTVCDelegate`, which in turn delegates to `NearbyBusinessesDataSource`.

jacobvanorder commented on Business Finder/MapViewController.h:
```diff
@@ -0,0 +1,21 @@
+@interface MapViewController : UIViewController<MKMapViewDelegate>
+-(void)setBusiness:(Business *)business;
+@property (weak, nonatomic) IBOutlet MKMapView *mapView;
+@property (strong, nonatomic) MapController *controller;
```
You have a double-strong relationship between `MapController` and `MapViewController`. I'd consider making one weak so you avoid retain loop.

notthatmj replied:
Fixed. `MapController` now keeps a weak reference to `MapViewController`.

jacobvanorder commented on Business Finder/MapViewController.m:
```diff
@@ -0,0 +1,77 @@
+    MKDirections *directions = [[MKDirections alloc] initWithRequest:directionsRequest];
+    [directions calculateDirectionsWithCompletionHandler:^(MKDirectionsResponse * _Nullable response,
+                                                           NSError * _Nullable error) {
+        if (!error && response.routes != nil && response.routes.count>0) {
```
spaces around `>`

notthatmj replied:
Fixed.

jacobvanorder commented on Business Finder/Model.h:
```diff
@@ -0,0 +1,28 @@
+- (void) modelDidFailWithError:(NSError *)error;
+@end
+
+@interface Model : NSObject<LocationGatewayDelegate,FourSquareGatewayDelegate>
```
Space after comma.

notthatmj replied:
Fixed.

jacobvanorder commented on Business Finder/NearbyBusinessesDataSource.h:
```diff
@@ -8,15 +8,15 @@
 
-@interface NearbyBusinessesDataSource : NSObject <UITableViewDataSource,BusinessesDataControllerDelegate>
-@property (strong,nonatomic) BusinessesDataController* businessesDataController;
+@interface NearbyBusinessesDataSource : NSObject <UITableViewDataSource,ModelObserving>
```
Space after comma.

notthatmj replied:
Fixed.

jacobvanorder commented on Business Finder/NearbyBusinessesTableViewController.m:
```diff
@@ -23,7 +25,30 @@ - (void)viewDidLoad {
+    [self.refreshControl beginRefreshing];
+    CGFloat offset = self.tableView.contentOffset.y-self.refreshControl.frame.size.height;
+    [self.tableView setContentOffset:CGPointMake(0, offset) animated:YES];
     [self.delegate startInitialLoad];
```
No `[super viewDidAppear:animated];` call

notthatmj replied:
Fixed.

jacobvanorder commented on Fastest Tests/BusinessAnnotationTests.m:
```diff
@@ -0,0 +1,29 @@
+- (void)setUp {
+    [super setUp];
+    // Put setup code here. This method is called before the invocation of each test method in the class.
+}
```
Empty method.

notthatmj replied:
Fixed.

jacobvanorder commented on Fastest Tests/MapControllerTests.m:
```diff
@@ -0,0 +1,72 @@
+
+- (void)setUp {
+    [super setUp];
+}
```
Empty method.

notthatmj replied:
Fixed.

jacobvanorder reviewed (commented):
Initial pass done

notthatmj committed:
Refactor

Remove `business` property of `BusinessCell`, add properties to replace the information it carried, and refactor related classes to accommodate this change.

notthatmj committed:
Fix strong reference cycle

notthatmj committed:
Remove empty methods and fix formatting

notthatmj committed:
Refactor

Remove `model` property from `AppDelegate`. Add `userLatitude` and
`userLongitude` properties to `MapViewController`, and modify
`NearbyBusinessesTableViewController` to configure these properties on the
destination view controller in its `prepareForSegue:` method.

notthatmj committed:
Remove `model` property from `AppDelegate`

notthatmj committed:
Add super call
